mutornadomon/external_interfaces/http_endpoints.py

```diff
@@ -23,95 +24,161 @@ def LOCALHOST(request):
-                }),
-            ])
+            }),
+        ])
 
-        self.tracer_app.listen(server_port)
-        logger.info('MuTornadoMon Tracer Listening on port %s', server_port)
+        # If listening is started directly then IOLoop started by service will
+        # cause issue resulting in high CPU usage. So start listening after
+        # IOLoop is started by the service
+        io_loop = IOLoop.current(instance=False)
+        if io_loop is None:
+            logger.error('Cannot initialize Mutornadomon without IOLoop')
+        else:
+            io_loop.add_callback(self.start_listen)
 
-    def start(self):
-        pass
+    def start_listen(self):
+        self.profiler_app.listen(self.server_port)
+        logger.info('MuTornadoMon Profiler Listening on port %s',
+                    self.server_port)
 
     def stop(self):
         pass
 
 
 class TornadoStatsHandler(tornado.web.RequestHandler):
     """
-    Start & return the Tornado IOLoop stack trace.
-    Tracing will be started when the url end point is hit & stopped after
-    waittime or default trace collection time expires
+    Profile Tornado IOLoop.
+    Profiler will be started when the url end point is hit & stopped after
+    profiletime or default profile collection time expires.
+    waittime starts the profiling periodically, profiling is done for the
+    duration of profiletime after waiting for a period of waittime.
     Params for the url are
-    :param sortby: specifies how the traces will be sorted
+    :param sortby: specifies how the profiling data will be sorted
         (ex: tottime or cumtime)
-    :param waittime: specifies how long the traces will be collected (msec)
-
-    ex: curl "localhost:5951/?sortby=cumtime&&waittime=4000"
+    :param profiletime: specifies how long profiling will be done (msec)
+    :param waittime: specifies how long to wait when profiling periodically
+    ex: curl "localhost:5951/profiler?sortby=cumtime&&profiletime=4000"
+    ex: curl "localhost:5951/profiler?profiletime=200&&waittime=10000"
     """
 
     def initialize(self, monitor, request_filter):
         self.monitor = monitor
         self.request_filter = request_filter
+        self.monitor.stop_profiler = False
 
     def prepare(self):
         if not self.request_filter(self.request):
             self.send_error(403)
 
-    @gen.coroutine
-    def get(self):
-        valid_sortby = ['calls', 'cumulative', 'cumtime', 'file', 'filename', 'module',
-                        'ncalls', 'pcalls', 'line', 'name', 'nfl', 'stdname', 'time', 'tottime']
+    def print_profile_data(self, sortby, wait_time):
+        # Stats fails if there is no profile data collected
+        try:
+            strm = StringIO()
+            ps = pstats.Stats(self.monitor.profiler, stream=strm)
+        except TypeError:
+            self.write("No profiling data collected")
+            return
+
+        if ps is not None:
```

Comment:
You need to init `ps = None` up above or else this will throw exception if it didn't get set within the try block above.